### PR TITLE
Backport small (low-risk) fixes from master to Copter 4.0.x stable

### PR DESCRIPTION
**Pull request by amilcarlucas**
These are the commits that did not get merged from #15151

### CODE REVIEW

**Review by tridge — APPROVED** [review]

**Comment by tridge** [comment]
looks good to me, needs check by @rmackay9

**Comment by amilcarlucas** [comment]
Rebased on top of 4.0.5-rc1.
@rmackay9 can you include this in 4.0.5-rc2 or 4.0.6-rc1 ?

**Comment by rmackay9** [comment]
Ok, leaving this until after the Copter-4.0.5 release goes out.

File: libraries/APM_Control/AR_AttitudeControl.cpp [hunk 1]
```diff
@@ -68,18 +68,24 @@ const AP_Param::GroupInfo AR_AttitudeControl::var_info[] = {
     // @Param: _STR_RAT_FLTT
     // @DisplayName: Steering control Target filter frequency in Hz
     // @Description: Target filter frequency in Hz
+    // @Range: 0.000 100.000
+    // @Increment: 0.1
```
Comment (muramura) [inline libraries/APM_Control/AR_AttitudeControl.cpp:72]: The number of decimal places is not correct.
I would suggest matching the number of digits.
The same goes for the other items.

**Comment by amilcarlucas** [comment]
@oceanalan yes this only changes the health flag

**Comment by rmackay9** [comment]
No reason to include the Rover change..

File: APMrover2/ekf_check.cpp [hunk 2]
```diff
@@ -153,7 +153,6 @@ void Rover::failsafe_ekf_event()
     failsafe.ekf = true;
     AP::logger().Write_Error(LogErrorSubsystem::FAILSAFE_EKFINAV,
                              LogErrorCode::FAILSAFE_OCCURRED);
-    gcs().send_text(MAV_SEVERITY_CRITICAL,"EKF failsafe!");
```
Comment (tridge) [inline APMrover2/ekf_check.cpp:156]: remove all the rover changes please

File: ArduPlane/altitude.cpp [hunk 3]
```diff
@@ -676,7 +676,9 @@ void Plane::rangefinder_height_update(void)
         if (now - rangefinder_state.last_correction_time_ms > 5000) {
             rangefinder_state.correction = correction;
             rangefinder_state.initial_correction = correction;
-            landing.set_initial_slope();
+            if (g.rangefinder_landing) {
```
Comment (tridge) [inline ArduPlane/altitude.cpp:679]: these are already in plane stable

File: ArduPlane/servos.cpp [hunk 4]
```diff
@@ -169,10 +169,6 @@ void Plane::channel_function_mixer(SRV_Channel::Aux_servo_function_t func1_in, S
  */
 void Plane::flaperon_update(int8_t flap_percent)
 {
-    if (!SRV_Channels::function_assigned(SRV_Channel::k_flaperon_left) &&
-        !SRV_Channels::function_assigned(SRV_Channel::k_flaperon_right)) {
```
Comment (tridge) [inline ArduPlane/servos.cpp:173]: no backport please

File: libraries/AC_Avoidance/AC_Avoid.cpp [hunk 5]
```diff
@@ -35,7 +35,7 @@ const AP_Param::GroupInfo AC_Avoid::var_info[] = {
     // @Values: 0:None,1:UseFence,2:UseProximitySensor,3:UseFence and UseProximitySensor,4:UseBeaconFence,7:All
     // @Bitmask: 0:UseFence,1:UseProximitySensor,2:UseBeaconFence
     // @User: Standard
-    AP_GROUPINFO("ENABLE", 1,  AC_Avoid, _enabled, AC_AVOID_DEFAULT),
+    AP_GROUPINFO_FLAGS("ENABLE", 1,  AC_Avoid, _enabled, AC_AVOID_DEFAULT, AP_PARAM_FLAG_ENABLE),
```
Comment (tridge) [inline libraries/AC_Avoidance/AC_Avoid.cpp:38]: not a good one for backport

File: libraries/AP_GPS/AP_GPS.cpp [hunk 6]
```diff
@@ -1244,16 +1244,17 @@ void AP_GPS::handle_gps_rtcm_fragment(uint8_t flags, const uint8_t *data, uint8_
         }
     }
 
-    uint8_t fragment = (flags >> 1U) & 0x03;
-    uint8_t sequence = (flags >> 3U) & 0x1F;
+    const uint8_t fragment = (flags >> 1U) & 0x03;
```
Comment (tridge) [inline libraries/AP_GPS/AP_GPS.cpp:1247]: not a good one for backport

File: libraries/AP_GPS/AP_GPS_UBLOX.cpp [hunk 7]
```diff
@@ -1176,15 +1176,24 @@ AP_GPS_UBLOX::_parse_gps(void)
                                              state.instance + 1,
                                              _version.hwVersion,
                                              _version.swVersion);
-            // check for F9. The F9 does not respond to SVINFO, so we need to use MON_VER
-            // for hardware generation
+            // check for F9 and M9. The F9 does not respond to SVINFO,
+            // so we need to use MON_VER for hardware generation
```
Comment (tridge) [inline libraries/AP_GPS/AP_GPS_UBLOX.cpp:1180]: not a good one for backport

File: libraries/AP_GPS/AP_GPS_UBLOX.h [hunk 8]
```diff
@@ -349,9 +349,11 @@ class AP_GPS_UBLOX : public AP_GPS_Backend
         uint32_t s_acc; 
         uint32_t head_acc; 
         uint16_t p_dop; 
-        uint8_t reserved1[6]; 
-        uint32_t headVeh;
-        uint8_t reserved2[4]; 
+        uint8_t flags3;
+        uint8_t reserved1[5];
```
Comment (tridge) [inline libraries/AP_GPS/AP_GPS_UBLOX.h:353]: not a good one for backport

File: libraries/AP_L1_Control/AP_L1_Control.cpp [hunk 9]
```diff
@@ -205,9 +205,13 @@ void AP_L1_Control::update_waypoint(const struct Location &prev_WP, const struct
 
     uint32_t now = AP_HAL::micros();
     float dt = (now - _last_update_waypoint_us) * 1.0e-6f;
+    if (dt > 1) {
```
Comment (tridge) [inline libraries/AP_L1_Control/AP_L1_Control.cpp:208]: not a good one for backport

File: libraries/AP_Math/AP_Math.cpp [hunk 10]
```diff
@@ -265,9 +265,11 @@ T constrain_value(const T amt, const T low, const T high)
     // the check for NaN as a float prevents propagation of floating point
     // errors through any function that uses constrain_value(). The normal
     // float semantics already handle -Inf and +Inf
-    if (isnan(amt)) {
-        AP::internalerror().error(AP_InternalError::error_t::constraining_nan);
-        return (low + high) / 2;
+    if (std::is_floating_point<T>::value) {
```
Comment (tridge) [inline libraries/AP_Math/AP_Math.cpp:268]: not a good one for backport

File: libraries/AP_NavEKF2/AP_NavEKF2_Measurements.cpp [hunk 11]
```diff
@@ -52,6 +52,8 @@ void NavEKF2_core::readRangeFinder(void)
                 }
                 storedRngMeasTime_ms[sensorIndex][rngMeasIndex[sensorIndex]] = imuSampleTime_ms - 25;
                 storedRngMeas[sensorIndex][rngMeasIndex[sensorIndex]] = sensor->distance_cm() * 0.01f;
+            } else {
+                continue;
```
Comment (tridge) [inline libraries/AP_NavEKF2/AP_NavEKF2_Measurements.cpp:56]: not a good one for backport

File: libraries/AP_RCProtocol/AP_RCProtocol_FPort.cpp [hunk 12]
```diff
@@ -124,8 +124,8 @@ void AP_RCProtocol_FPort::decode_control(const FPort_Frame &frame)
 
     bool failsafe = ((frame.control.flags & (1 << FLAGS_FAILSAFE_BIT)) != 0);
 
-    // we scale rssi by 2x to make it match the value displayed in OpenTX
-    const uint8_t scaled_rssi = MIN(frame.control.rssi*2, 255);
+    // fport rssi 0-50, ardupilot rssi 0-255, scale factor 255/50=5.1
```
Comment (tridge) [inline libraries/AP_RCProtocol/AP_RCProtocol_FPort.cpp:127]: not a good one for backport

File: libraries/GCS_MAVLink/GCS.cpp [hunk 13]
```diff
@@ -193,6 +195,16 @@ void GCS::update_sensor_status_flags()
         }
     }
 
+    // give GCS status of prearm checks. This is enabled if any arming checks are enabled.
```
Comment (tridge) [inline libraries/GCS_MAVLink/GCS.cpp:198]: not a good one for backport

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 14]
```diff
@@ -1050,7 +1051,15 @@ void GCS_MAVLINK::update_send()
                 if (!do_try_send_message(deferred_message[next].id)) {
                     break;
                 }
-                deferred_message[next].last_sent_ms += deferred_message[next].interval_ms;
+                // we try to keep output on a regular clock to avoid
```
Comment (tridge) [inline libraries/GCS_MAVLink/GCS_Common.cpp:1054]: not a good one for backport

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 15]
```diff
@@ -1672,6 +1688,14 @@ void GCS_MAVLINK::send_scaled_pressure_instance(uint8_t instance, void (*send_fn
     if (airspeed != nullptr &&
         airspeed->enabled(instance)) {
         press_diff = airspeed->get_differential_pressure(instance) * 0.01f;
+        float temp;
+        if (airspeed->get_temperature(instance,temp)) {
```
Comment (tridge) [inline libraries/GCS_MAVLink/GCS_Common.cpp:1692]: not a good one for backport

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 16]
```diff
@@ -2103,6 +2127,11 @@ void GCS_MAVLINK::send_autopilot_version() const
     uint16_t product_id = 0;
     uint64_t uid = 0;
     uint8_t  uid2[MAVLINK_MSG_AUTOPILOT_VERSION_FIELD_UID2_LEN] = {0};
+
+    uint8_t uid_len = sizeof(uid2); // taken as reference and modified
+                                    // by following call:
+    hal.util->get_system_id_unformatted(uid2, uid_len);
```
Comment (tridge) [inline libraries/GCS_MAVLink/GCS_Common.cpp:2133]: not a good one for backport

File: libraries/GCS_MAVLink/MAVLink_routing.cpp [hunk 17]
```diff
@@ -172,7 +172,8 @@ bool MAVLink_routing::check_and_forward(mavlink_channel_t in_channel, const mavl
         }
     }
 
-    if (!forwarded && match_system) {
+    if ((!forwarded && match_system) ||
+        broadcast_system) {
```
Comment (tridge) [inline libraries/GCS_MAVLink/MAVLink_routing.cpp:176]: not a good one for backport

File: libraries/GCS_MAVLink/MAVLink_routing.cpp [hunk 18]
```diff
@@ -261,9 +262,20 @@ bool MAVLink_routing::find_by_mavtype(uint8_t mavtype, uint8_t &sysid, uint8_t &
...
 {
     uint8_t i;
-    if (msg.sysid == 0 ||
-        (msg.sysid == mavlink_system.sysid &&
-         msg.compid == mavlink_system.compid)) {
+    if (msg.sysid == 0) {
+        // don't learn routes to the broadcast system
+        return;
+    }
+    if (msg.sysid == mavlink_system.sysid &&
```
Comment (tridge) [inline libraries/GCS_MAVLink/MAVLink_routing.cpp:269]: not a good one for backport

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_NavEKF2/AP_NavEKF2_Measurements.cpp (hunk 11 above)
Comment (amilcarlucas) [reply libraries/AP_NavEKF2/AP_NavEKF2_Measurements.cpp:56]: This was already approved above, and fixes a safety bug

File: libraries/AP_RCProtocol/AP_RCProtocol_FPort.cpp (hunk 12 above)
Comment (amilcarlucas) [reply libraries/AP_RCProtocol/AP_RCProtocol_FPort.cpp:127]: This is a nice small non intrusive change.... why not let it in?

File: libraries/GCS_MAVLink/GCS.cpp (hunk 13 above)
Comment (amilcarlucas) [reply libraries/GCS_MAVLink/GCS.cpp:198]: Already approved above and at least 4 users asked for this.

File: libraries/APM_Control/AR_AttitudeControl.cpp (hunk 1 above)
Comment (amilcarlucas) [reply libraries/APM_Control/AR_AttitudeControl.cpp:72]: This would need to be also fixed in master, before I could backport. That would be another PR

**Comment by amilcarlucas** [comment]
@tridge Can you comment on my comments above?

**Comment by amilcarlucas** [comment]
Can someone re-trigger a build?

**Comment by peterbarker** [comment]
> Can someone re-trigger a build?

That's not going to help; fmuv2 is overflowing.

**Comment by amilcarlucas** [comment]
Can we ignore that? after all this is a copter release.

**Comment by amilcarlucas** [comment]
I moved all the interesting commits to other PRs. This PR only contains refused commits now, so closing it.

**Pull request closed without merging**